fix(mobile): do not removed not backup asset when selecting the correspond options

yashrajjain726 opened the pull request:
Fixed #11428

yashrajjain726 committed:
fixed the local ids selecting issue

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 committed:
Merge branch 'main' into fix/11428

shenlong-tanwen commented on mobile/lib/widgets/asset_grid/multiselect_grid.dart:
```diff
@@ -204,11 +204,26 @@ class MultiselectGrid extends HookConsumerWidget {
+        }
+
+        // Proceed with deletion of filtered assets
         final isDeleted = await ref
             .read(assetProvider.notifier)
             .deleteLocalOnlyAssets(localIds, onlyBackedUp: onlyBackedUp);
```
The `onlyBackup` parameter is already handled in `deleteLocalOnlyAssets` method. The fix to this issue should ideally be made inside the method.

yashrajjain726 committed:
code: updated impl inside deleteLocalOnlyAssets

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 commented:
@shenlong-tanwen updated the implementation inside the deleteLocalOnlyAssets func:

yashrajjain726 committed:
fix: used png instead of jpg to maintain picture quality

yashrajjain726 committed:
Revert "fix: used png instead of jpg to maintain picture quality"

This reverts commit 04f2ed54e40b454a057f97872836d71af50b06de.

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 commented:
@shenlong-tanwen  Any update on review ?

yashrajjain726 committed:
Merge branch 'main' into fix/11428

shenlong-tanwen commented on mobile/lib/providers/asset.provider.dart:
```diff
@@ -78,27 +80,48 @@ class AssetNotifier extends StateNotifier<bool> {
+
+      // Continue with deletion of the filtered assets
       final assets = onlyBackedUp
-          ? deleteAssets.where((e) => e.storage == AssetState.merged)
-          : deleteAssets;
```
This section of code already filters the total list passed based on the backed up status of the asset. i.e, `e.storage == AssetState.merged` essentially filters only those that are already backed up. You added the same filtering more explicitly before this thus making this iteration useless

shenlong-tanwen commented on mobile/lib/providers/asset.provider.dart:
```diff
@@ -78,27 +80,48 @@ class AssetNotifier extends StateNotifier<bool> {
       final localDeleted = await _deleteLocalAssets(assets);
       if (localDeleted.isNotEmpty) {
-        final localOnlyIds = deleteAssets
+        final localOnlyIds = assetsToDelete
             .where((e) => e.storage == AssetState.local)
             .map((e) => e.id)
             .toList();
         // Update merged assets to remote only
-        final mergedAssets =
-            deleteAssets.where((e) => e.storage == AssetState.merged).map((e) {
```
This portion of the code is actually where the logical error is. So if we do not need the toast message if there are not assets to delete, the entirety of the change would only be switching the list from `deleteAssets` to `assets` in the current implementation

Basically, the issue is that, we filter the list of assets and get the final list to delete but when making the changes to the local DB, we filter the assets again from the original list passed and not the filtered list

shenlong-tanwen commented on mobile/lib/providers/asset.provider.dart:
```diff
@@ -78,27 +80,48 @@ class AssetNotifier extends StateNotifier<bool> {
+          .where((e) => e.isLocal && (!onlyBackedUp || e.isRemote))
+          .toList();
+
+      if (assetsToDelete.isEmpty) {
+        // No assets to delete, exit early
+        ImmichToast.show(
+          context: context,
+          msg: 'No backed-up assets selected for deletion.',
+          gravity: ToastGravity.BOTTOM,
+        );
```
Toast handling can be made in the widget and not the provider. In general, we shouldn't be passing `BuildContext` inside business logic unless we have no other option.

shenlong-tanwen commented on mobile/lib/widgets/asset_grid/multiselect_grid.dart:
```diff
@@ -204,10 +204,11 @@ class MultiselectGrid extends HookConsumerWidget {
       processing.value = true;
       try {
         final localIds = selection.value.where((a) => a.isLocal).toList();
```
The best way to handle this is to use the `onlyBackedUp` flag to filter assets directly here instead of having the filtering logic in the provider. Having the logic here will also allow you to show the toast when there are no assets to delete.

shenlong-tanwen reviewed (changes requested):
You could do one of the following:

i) You could use your previous implementation of having the filtering logic in the widget and remove all the filtering logic in the provider as it is redundant and has errors
ii) Fix the logical issue in the provider and remove your filtering logic and toast handling in the widget as the assets are already filtered in the provider

Also fix any issues from the dart analyzer as well

Apologies for not being more clearer with the previous comment. Thanks for working on this.

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 committed:
fix: update logic from code-review perspective

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 commented:
Thanks @shenlong-tanwen , I have updated the logic as per requirement. And, also removed UI logic from the provider code, as you mentioned and have implemented it from code perpective

shenlong-tanwen commented on mobile/lib/widgets/asset_grid/multiselect_grid.dart:
```diff
@@ -203,19 +203,33 @@ class MultiselectGrid extends HookConsumerWidget {
+
+          // Reset the selection
           selectionEnabledHook.value = false;
+
+          // Trigger a refresh to update the UI after deletion
+          await ref.read(assetProvider.notifier).getAllAsset();
```
Why is this needed?

shenlong-tanwen reviewed (commented):
Kindly fix the dart analysis issues. i.e, there are unused imports in the provider file

yashrajjain726 committed:
refractor (mobile) : Dart fix applied

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 committed:
Merge branch 'main' into fix/11428

shenlong-tanwen commented:
@yashrajjain726 The changes are good.Can you respond to the only remaining comment on why the refresh is needed?

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 commented:
I have replied to it long time ago, please check.

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 committed:
fix (mobile) : Updated multi grid as per requirement

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 commented:
Hi @shenlong-tanwen , Can you please check again, I have updated as required.

shenlong-tanwen reviewed (approved):
Looks great! Thank you

yashrajjain726 committed:
Merge branch 'main' into fix/11428

yashrajjain726 commented:
@alextran1502 Any update on the review ?

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into fix/11428
